### PR TITLE
add type check for io

### PR DESCRIPTION
**Pull request by LHT129**

### CODE REVIEW

File: src/io/basic_io.h [hunk 1]
```diff
@@ -31,42 +31,83 @@ class BasicIO {
...
+        }
     }
 
     inline bool
     Read(uint64_t size, uint64_t offset, uint8_t* data) const {
-        return cast().ReadImpl(size, offset, data);
+        if constexpr (std::is_member_function_pointer_v<decltype(&IOTmpl::ReadImpl)>) {
+            return cast().ReadImpl(size, offset, data);
+        } else {
+            throw std::runtime_error(
```
Comment (jiaweizone) [inline src/io/basic_io.h:62]: add test code for this else clause

Comment (LHT129) [reply src/io/basic_io.h:62]: fixed

**Review by inabao — APPROVED** [review]
LGTM

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**